### PR TITLE
Initial code cleanup for rewrite

### PR DESCRIPTION
**Pull request by qpresley**
### Description
Starting a re-write based on new versions of ibm_db which includes updated connection pooling and improvements in async operation.  This also is based on updates to loopback-connector and loopback-ibmdb to provide more consistency and code standards across the connectors.

- [x ] New tests added or existing tests modified to cover all changes
- [x ] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: lib/db2i.js [hunk 1]
```diff
@@ -95,283 +95,150 @@ DB2.prototype.create = function(model, data, options, callback) {
...
-    connection.query(countStmt, function(err, countData) {
-      debug('DB2.prototype.update (data): sql %j, countData=%j\n',
-            countStmt, countData);
-
+  var id = self.idName(model);
+  var deleteStmt = self.buildDelete(model, where, options);
+  var selectStmt = new ParameterizedSQL('SELECT ' + self.escapeName(id) +
+                                        ' FROM ' + tableName);
+  selectStmt.merge(self.buildWhere(model, where));
+  // selectStmt.merge(' FOR UPDATE');
```
Comment (0candy) [inline lib/db2i.js]: This line can be removed.

File: lib/migration.js [hunk 2]
```diff
@@ -80,39 +68,39 @@ module.exports = function(DB2iSeries) {
...
-  DB2iSeries.prototype.getTableStatus = function(model, cb) {
-    process.nextTick(function() {
-      return cb(Error(g.f('{{getTableStatus()}} is not currently supported.')));
-    });
-  };
+//   DB2iSeries.prototype.getTableStatus = function(model, cb) {
+//     process.nextTick(function() {
+//       return cb(Error(g.f('{{getTableStatus()}} is not currently supported.')));
+//     });
+//   };
```
Comment (0candy) [inline lib/migration.js]: Do we need all the commented out methods in this file or can they be removed?

File: test/init.js [hunk 3]
```diff
@@ -15,7 +15,7 @@ var config = {
   hostname: process.env.DB2I_HOSTNAME || 'localhost',
   port: process.env.DB2I_PORTNUM || 60000,
   database: process.env.DB2I_DATABASE || 'testdb',
-  schema: process.env.DB2I_SCHEMA || 'STRONGLOOP',
+  // schema: process.env.DB2I_SCHEMA || 'STRONGLOOP',
```
Comment (0candy) [inline test/init.js]: This line can be removed.

File: lib/migration.js (hunk 2 above)
Comment (qpresley) [reply lib/migration.js]: Kind of hard to tell now that I look at it in more detail.  I was expecting autoupdate among other functions to be common and to be in loopback-ibmdb but they aren't.  I'm going to uncomment all these.  I don't think there are migration tests for this connector yet so it should be fine to ship with this but if I can get it working quickly I'll likely push changes to this as well as IBMDB and add the migration tests.

Comment (qpresley) [reply lib/migration.js]: Had a quick look.  I'm going to state that migration is not supported at this time for this connector as it is going to require writing these functions from scratch as the system catalogs are different than those from the other IBM databases.  We'll have to do a minor update to add migration/discovery later.

File: lib/transaction.js [hunk 4]
```diff
@@ -32,16 +32,16 @@ var mapIsolationLevel = function(isolationLevelString) {
 };
 
 /*!
- * @param {DB2} DB2 connector class
+ * @param {DB2iSeries} DB2 iSeries connector class
```
Comment (b-admike) [inline lib/transaction.js:35]: JSdoc for `db2i` parameter here

File: lib/migration.js [hunk 5]
```diff
@@ -13,106 +13,76 @@ var g = require('./globalize');
...
-      return cb(Error(g.f('{{alterTable()}} is not currently supported.')));
-    });
-  };
-
-  DB2iSeries.prototype.getTableStatus = function(model, cb) {
-    process.nextTick(function() {
-      return cb(Error(g.f('{{getTableStatus()}} is not currently supported.')));
-    });
-  };
+  // DB2iSeries.prototype.isActual = function(models, cb) {
```
Comment (b-admike) [inline lib/migration.js]: remove if dead code, otherwise you can explain why it is commented out

File: lib/migration.js [hunk 6]
```diff
@@ -13,106 +13,76 @@ var g = require('./globalize');
...
+    var sql = 'SELECT INDEX_NAME as INDNAME ' +
+              'FROM QSYS2.SYSINDEXES ' +
+              'WHERE TRIM(TABLE_NAME) = \'' + self.table(model) + '\' ' +
+              'AND TRIM(TABLE_SCHEMA) = \'' + self.schema.toUpperCase() + '\'';
+
+    this.execute(sql, function(err, indexes) {
+      if (err) {
+        return cb(err);
+      } else {
+        cb(err, indexes);
```
Comment (b-admike) [inline lib/migration.js]: `return cb(err,indexes);`

File: lib/migration.js [hunk 7]
```diff
@@ -13,106 +13,76 @@ var g = require('./globalize');
...
+              self.table(model) + '\' ' +
+              'AND TRIM(TABLE_SCHEMA) LIKE \'' +
+              self.schema.toUpperCase() + '\'' +
+              ' ORDER BY COLNO';
+
+    this.execute(sql, function(err, fields) {
+      if (err) {
+        return cb(err);
+      } else {
+        cb(err, fields);
```
Comment (b-admike) [inline lib/migration.js]: `return cb(err, fields);`

File: lib/migration.js [hunk 8]
```diff
@@ -13,106 +13,76 @@ var g = require('./globalize');
...
+  //     'currently supported.')));
+  //   });
+  // };
 
   DB2iSeries.prototype.showFields = function(model, cb) {
-    process.nextTick(function() {
-      return cb(Error(g.f('{{showFields()}} is not currently supported.')));
+    var self = this;
+    var sql = 'SELECT COLUMN_NAME AS NAME, DATA_TYPE AS DATATYPE, ' +
+              'ORDINAL_POSITION AS COLNO,' +
```
Comment (b-admike) [inline lib/migration.js]: need space after comma? (following the syntax from previous line)

File: lib/migration.js [hunk 9]
```diff
@@ -13,106 +13,76 @@ var g = require('./globalize');
...
-  };
-
-  DB2iSeries.prototype.applySqlChanges =
-  function(model, pendingChanges, cb) {
-    process.nextTick(function() {
-      return cb(Error(g.f('{{applySqlChanges()}} is not ' +
-      'currently supported.')));
-    });
-  };
+  // DB2iSeries.prototype.searchForPropertyInActual =
```
Comment (b-admike) [inline lib/migration.js]: remove if dead code, otherwise comment as a block and specify why

File: lib/discovery.js [hunk 10]
```diff
@@ -10,124 +10,125 @@ var g = require('./globalize');
 module.exports = mixinDiscovery;
 
 /**
-* @param {DB2} DB2 connector class
-* @param {Object} db2
+* @param {DB2iSeries} DB2iSeries connector class
+* @param {Object} db2i
```
Comment (b-admike) [inline lib/discovery.js]: what does the `db2i` object represent here?

File: lib/db2i.js [hunk 11]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
+      } else {
+        executeTransaction(conn, function(err, data) {
+          if (err) {
+            self.rollback(conn, function(err) {});
+            conn.close(function(err) {});
+            return callback(err);
+          }
+
+          self.commit(conn, function(err) {});
+          callback(null, data);
```
Comment (b-admike) [inline lib/db2i.js]: `return callback(null, data);`

File: lib/db2i.js [hunk 12]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
   } else {
-    self.beginTransaction(Transaction.READ_COMMITTED, function(err, conn) {
+    self.beginTransaction(Transaction.REPEATABLE_READ, function(err, conn) {
       if (err) {
-        return callback && callback(err);
+        return callback(err);
+      } else {
+        executeTransaction(conn, function(err, data) {
+          if (err) {
+            self.rollback(conn, function(err) {});
```
Comment (b-admike) [inline lib/db2i.js]: handle `err` here

File: lib/db2i.js [hunk 13]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
+            conn.close(function(err) {});
```
Comment (b-admike) [inline lib/db2i.js]: handle `err` here

File: lib/db2i.js [hunk 14]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
-      debug('DB2.prototype.destroyAll (data)', stmt.sql, stmt.params);
-      connection.query(stmt, function(err, sData) {
-        if (err) {
-          debug('DB2.prototype.destroyAll (ERR): %j', err);
-          return cb(err);
-        }
-        data.count = countData[0]['CNT'];
-        cb(null, data);
-      });
+      callback(null, data);
```
Comment (b-admike) [inline lib/db2i.js]: `return callback(null, data);`

File: lib/db2i.js [hunk 15]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
+  var id = self.idName(model);
+  var deleteStmt = self.buildDelete(model, where, options);
+  var selectStmt = new ParameterizedSQL('SELECT ' + self.escapeName(id) +
+                                        ' FROM ' + tableName);
+  selectStmt.merge(self.buildWhere(model, where));
+  self.parameterize(selectStmt);
+
+  var executeTransaction = function(connection, cb) {
+    connection.query(selectStmt, function(err, selectData) {
+      debug('DB2iSeries.prototype.destroyAll stmt: %j data: %j', selectStmt,
```
Comment (b-admike) [inline lib/db2i.js:193]: I would suggest that you have this debug statement after handling `err`

File: lib/db2i.js [hunk 16]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
+        return callback(err);
+      } else {
+        executeTransaction(conn, function(err, retVal) {
+          if (err) {
+            self.rollback(conn, function(err) {});
+            conn.close(function(err) {});
+            return callback(err);
+          }
+          self.commit(conn, function(err) {});
+          callback(null, {count: retVal});
```
Comment (b-admike) [inline lib/db2i.js]: `return callback(null, {count: retVal});`

File: lib/db2i.js [hunk 17]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
   } else {
-    self.beginTransaction(Transaction.READ_COMMITTED, function(err, conn) {
+    self.beginTransaction(Transaction.REPEATABLE_READ, function(err, conn) {
       if (err) {
-        return callback && callback(err);
+        return callback(err);
+      } else {
+        executeTransaction(conn, function(err, retVal) {
+          if (err) {
+            self.rollback(conn, function(err) {});
```
Comment (b-admike) [inline lib/db2i.js]: do we care about the `err` returned here? if so, we should handle it

File: lib/db2i.js [hunk 18]
```diff
@@ -95,283 +94,149 @@ DB2.prototype.create = function(model, data, options, callback) {
...
-          return callback && callback(err);
-        } else {
-          return callback && callback(null, data, meta);
-        }
-      });
+    executeTransaction(options.transaction.connection, function(err, retVal) {
+      if (err) {
+        return callback(err);
+      }
+      callback(null, {count: retVal});
```
Comment (b-admike) [inline lib/db2i.js]: `return callback(null, {count: retVal});`

**Review by b-admike — CHANGES_REQUESTED** [review]

File: lib/db2i.js (hunk 17 above)
Comment (qpresley) [reply lib/db2i.js]: No, it isn't really relevant as we can't really do anything about it.

Comment (qpresley) [reply lib/db2i.js]: I don't like the code order here though and am changing to have the conn.close inside the rollback callback.

File: lib/db2i.js (hunk 15 above)
Comment (qpresley) [reply lib/db2i.js:193]: The selectData happens to contain the Error array from the statement executed so having this hear gives me the benefit of being able to dump the error content in the debug statement.

Comment (b-admike) [reply lib/db2i.js:193]: k

**Pull request merged**